libsnark/zk_proof_systems/plonk/prover.tcc

```diff
@@ -0,0 +1,989 @@
+        srs.secret_powers_g1, W_zeta_omega);
+
+    round_five_out_t<ppT> round_five_out(
+        std::move(r_zeta),
+        std::move(W_zeta_at_secret),
+        std::move(W_zeta_omega_at_secret));
+
+    return round_five_out;
+}
+
+/// Prover compute SNARK proof
+///
+/// Pi ([a]_1, [b]_1, [c]_1, [z]_1,
+///     [t_lo]_1, [t_mi]_1, [t_hi]_1,
+///     \bar{a}, \bar{b}, \bar{c},
+///     \bar{S_sigma1}, \bar{S_sigma2}, \bar{z_w},
+///     [W_zeta]_1, [W_{zeta omega}]_1
+///     r_zeta)
+///
+/// \note in the reference Python implementation, r_zeta (the
+/// evaluation of the linearlization polynomial r(X) at zeta from
+/// Prover round 5) is added to the pi-SNARK proof. In the paper this
+/// is omitted, which seems to make the proof shorter by 1 element at
+/// the epxense of a slightly heavier computation on the verifier's
+/// side. Here we follow the reference implementation to make sure we
+/// match the test values. TODO: once all test vectors are verified,
+/// we may remove r_zeta from the proof to be fully compliant with the
+/// paper.
+///
+/// Mapping code-to-paper quantities
+///
+/// \param W_polys_blinded_at_secret_g1[a, b, c]: [a]_1, [b]_1, [c]_1
+///        (from Round 1)
+/// \param z_poly_at_secret_g1: [z]_1 (from Round 2)
+/// \param t_poly_at_secret_g1[lo, mi, hi]: [t_lo]_1, [t_mi]_1,
+///        [t_hi]_1 (from Round 3)
+/// \param a_zeta, b_zeta, c_zeta, S_0_zeta, S_1_zeta,
+///        z_poly_xomega_zeta: \bar{a}, \bar{b}, \bar{c},
+///        \bar{S_sigma1}, \bar{S_sigma2}, \bar{z_w} (from Round 4)
+/// \param W_zeta_at_secret, W_zeta_omega_at_secret: [W_zeta]_1,
+///        [W_{zeta omega}]_1 (from Round 5)
+///
+/// INPUT
+/// \param[in] srs: structured reference string containing also
+///            circuit-specific information
+/// \param[in] witness: all internal values and public input
+///            corresponding to the given circuit
+/// \param[in] blind_scalars: random blinding scalars b1, b2, ..., b9
+///            used in prover rounds 1 and 2 (see Sect. 8.3, roumds
+///            1,2 [GWC19])
+/// \param[in] transcript_hash: hashes of the communication transcript
+///            after prover rounds 1,2,3,4,5. TODO: \attention
+///            currently the structure is used as an input initialized
+///            with hard-coded example values for debug purposes. In
+///            the long run it should be modified to be used as an
+///            output. More specifically, the hard-coded values should
+///            be overwritten with the actual transcript hashes
+///            produced after the respective rounds within \ref
+///            compute_proof
+///
+/// OUTPUT
+/// \param[out] proof: SNARK proof Pi (see above)
+template<typename ppT>
+plonk_proof<ppT> plonk_prover<ppT>::compute_proof(
+    const srs<ppT> &srs,
+    const std::vector<Field> &witness,
+    const std::vector<libff::Fr<ppT>> blind_scalars,
+    transcript_hash_t<ppT> &transcript_hash)
```

Comment:
Is this an input or an output?  In reality they must be computed during the proof generation process, using the intermediate data.

Reply:
Using a hash function to compute these may be one for the future, but even for now I'm a bit nervous about creating a data type `transcript_hash_t` that has them pre-populated before calling "compute_proof".

Reply:
We might want something like a `transcript_hasher` which can accept field and group elements as they are available, and then output a digest when required.
That's a bit of a bigger change, though.  So maybe we have to keep `transcript_hash_t` for now, and add this as an issue for later.

Reply:
Added this as a TODO as part of Issue https://github.com/clearmatics/libsnark/issues/52